plugin/oauth2/src/OAuth2.php

```diff
@@ -0,0 +1,382 @@
+    {
+        if (!is_string($key) || empty($key) || !count($data)) {
+            return $default;
+        }
+        $pos = strpos($key, '*');
+        if ($pos === false) {
+            $value = $this->getValueByKey($data, $key, null);
+
+            return is_null($value) ? [] : [$value];
+        }
+        $values = [];
+        $beginning = substr($key, 0, $pos);
+        $remaining = substr($key, $pos + 1);
+        $index = 0;
+        do {
+            $newValues = $this->getValuesByKey(
+                $data,
+                $beginning.$index.$remaining
+            );
+            $values = array_merge($values, $newValues);
+            $index++;
+        } while ($newValues);
+
+        return $values;
+    }
+
+    private function updateUser($userId, $response)
+    {
+        /**
+         * @var $user Chamilo\UserBundle\Entity\User
+         */
+        $user = UserManager::getRepository()->find($userId);
+        $user->setFirstname(
+            $this->getValueByKey($response, $this->get(
+                self::SETTING_RESPONSE_RESOURCE_OWNER_FIRSTNAME
+            ), $user->getFirstname())
+        );
+        $user->setLastname(
+            $this->getValueByKey($response, $this->get(
+                self::SETTING_RESPONSE_RESOURCE_OWNER_LASTNAME
+            ), $user->getLastname())
+        );
+        $user->setUserName(
+            $this->getValueByKey($response, $this->get(
+                self::SETTING_RESPONSE_RESOURCE_OWNER_USERNAME
+            ), $user->getUsername())
+        );
+        $user->setEmail(
+            $this->getValueByKey($response, $this->get(
+                self::SETTING_RESPONSE_RESOURCE_OWNER_EMAIL
+            ), $user->getEmail())
+        );
+        $user->setStatus(
+            $this->getValueByKey($response, $this->get(
+                self::SETTING_RESPONSE_RESOURCE_OWNER_STATUS
+            ), $user->getStatus())
+        );
+        $user->setAuthSource('oauth2');
+        $configFilePath = __DIR__.'/../config.php';
+        if (file_exists($configFilePath)) {
+            require_once $configFilePath;
+            $functionName = 'oauth2_update_user_from_resource_owner_details';
+            if (function_exists($functionName)) {
+                $functionName($response, $user);
+            }
+        }
+        UserManager::getManager()->updateUser($user);
+    }
+
+    /**
+     * Updates the Access URLs associated to a user
+     * according to the OAuth2 server response resource owner
+     * if multi-URL is enabled and SETTING_RESPONSE_RESOURCE_OWNER_URLS defined.
+     *
+     * @param $userId integer
+     * @param $response array
+     */
+    private function updateUserUrls($userId, $response)
+    {
+        if (api_is_multiple_url_enabled()) {
+            $key = $this->get(self::SETTING_RESPONSE_RESOURCE_OWNER_URLS);
+            if (!empty($key)) {
+                require_once __DIR__.'/../../../main/inc/lib/urlmanager.lib.php';
```

Comment:
"urlmanager.lib.php" should be already loaded by composer, so no need to do this require_once